### ISSUE TITLE
Failure with >= 100 histogram bins

### ISSUE THREAD

**Issue by dburov190** [mirkobunse/EarthMoversDistance.jl#3]
The following MWE has been consistently failing for me:
```julia
using EarthMoversDistance
earthmovers(rand(100), rand(100), (x,y) -> abs(x - y))
```

Is there some kind of limitation on the number of bins?

**Comment by mirkobunse** [comment]
Hi,

yes, indeed the number of bins is limited. Keep in mind that this Julia package is meant to wrap the original EMD implementation by Rubner et al. It is this original implementation which uses a maximum of 100 bins.

As far as I can tell, Rubner et al have limited the number of bins to not allocate too much memory without using it. In the file `deps/src/emd.h` you can alter the limit by changing the value of `MAX_SIG_SIZE` (line 20) and recompile the EMD library. Let me provide you with an alternative `deps/build.jl` which does so:

```julia
# Most of this code is the regular content of deps/build.jl

using Libdl

# where to get the C source files from
HEADER_URL         = "http://robotics.stanford.edu/~rubner/emd/emd.h"
IMPLEMENTATION_URL = "http://robotics.stanford.edu/~rubner/emd/emd.c"

# clean up
rm("src", force=true, recursive=true)

# download the source files
mkdir("src")
download(HEADER_URL,         "src/" * basename(HEADER_URL))
download(IMPLEMENTATION_URL, "src/" * basename(IMPLEMENTATION_URL))

# feature type should be float (funny -i'' works on both Linux and Mac)
run(`sed -i\'\' 's/typedef int feature_t;/typedef float feature_t;/g' src/emd.h`)


##################################################
# The following line sets the maximum number of bins to 1000
##################################################
run(`sed -i\'\' 's/MAX_SIG_SIZE   100/MAX_SIG_SIZE   1000/g' src/emd.h`)


# create a shared library: compile and link
libfile = "emd.$(Libdl.dlext)" # extension depends on OS
run(`gcc -Wall -fpic -shared -o $libfile src/emd.c`)

# check for successful creation of the library
if !isfile(libfile)
    error("The shared library $libfile was not created during the build")
end
```

**Caution:** When I tested the library with a bin limit of 1000, I frequently ran into segmentation faults, also when computing distances for histograms with less than 100 bins.

You can try to reduce the number of bins by aggregation. Or, you may try another algorithm. For histograms, there is the Minimum Distance of Pair Assignments [1]. In my Master's thesis [2], page 35, I described this algorithm in simpler terms.

[1] S. H. Cha and S.N. Srihari: On measuring the distance between histograms. In: Pattern Recognition, vol. 25, no. 6, pp.1355-1370, 2002.
[2]  M. Bunse: DSEA Rock-Solid - Regularization and Comparison with other Deconvolution Algorithms. TU Dortmund, 44221 Dortmund, Germany, 2018. Link to the PDF: https://www-ai.cs.uni-dortmund.de/auto?self=%24Publication_fh34l6f01s

**Comment by dburov190** [comment]
> Caution: When I tested the library with a bin limit of 1000, I frequently ran into segmentation faults, also when computing distances for histograms with less than 100 bins.

Hmm, seems like you're right, I caught seg faults on 90s and even 80s. Hm.... I guess I'll stick with `PyCall` and `scipy.stats.wasserstein_distance` for now 😬 Honestly, the code is very simple but I don't see the point in, basically, copy-pasting from `scipy` into some Julia package. `PyCall` only introduces slight overhead, based on what I've seen.

Thanks for a detailed answer though! I'm using Wasserstein distance as a metric to tell how far apart two empirical distributions (i.e. histograms) are. I basically don't care about the absolute values themselves, I will simply choose some threshold by trial and error, be it 0.1 or 100. Wasserstein is just better than K-L since it's symmetric, but other than that I don't have any preference really. Thanks for the links, I'm reading the Cha & Srihari paper.

PS I guess we can close it now? Not a real issue with this package.
